File: typescript/src/address/lookup/lookupV2ResultAddressFormatted.ts
import { RestApiAddressLookupV2ResultAddressFormatted } from "../../server/address/lookup/restApiAddressLookupV2ResultAddressFormatted";
import { LookupElectricityMeter, restApiResponseToLookupElectricityMeter } from "./lookupElectricityMeter";
import { LookupGasMeter, restApiResponseToLookupGasMeter } from "./lookupGasMeter";

export type LookupV2ResultAddressFormatted = {
    layoutName?: string;
    address?: {
        electricityMeters?: LookupElectricityMeter[],
        gasMeters?: LookupGasMeter[]
    };
}

export function restApiResponseToLookupAddressFormatted(response: RestApiAddressLookupV2ResultAddressFormatted): LookupV2ResultAddressFormatted {
    const result: LookupV2ResultAddressFormatted = {
        layoutName: response.layout_name,        
    };
    if (response.address) {
        result.address = {};
        if (response.address.electricity_meters) {
            result.address.electricityMeters = response.address.electricity_meters.map(e => restApiResponseToLookupElectricityMeter(e));
        }
        if (response.address.gas_meters) {
            result.address.gasMeters = response.address.gas_meters.map(g => restApiResponseToLookupGasMeter(g));
        }
    }
    return result;
}

File: csharp/DVSClient/DVSClient/address/lookup/LookupV2ResultAddressFormatted.cs
using DVSClient.Server.Address.Lookup;

namespace DVSClient.Address.Lookup
{
    public class LookupV2ResultAddressFormatted
    {
        public string? LayoutName { get; }
        public LookupV2ResultAddressFormattedAddress? Address { get; }

        public LookupV2ResultAddressFormatted(RestApiAddressLookupV2ResultAddressFormatted response)
        {
            LayoutName = response.LayoutName;
            if (response.Address != null) {
                Address = new LookupV2ResultAddressFormattedAddress(response.Address);
            }
        }
    }

    public class LookupV2ResultAddressFormattedAddress
    {
        public IEnumerable<LookupElectricityMeter>? ElectricityMeters { get; }
        public IEnumerable<LookupGasMeter>? GasMeters { get; }

        public LookupV2ResultAddressFormattedAddress(RestApiAddressLookupV2ResultAddressFormatted.AddressDetails addressDetails)
        {
            ElectricityMeters = addressDetails.ElectricityMeters != null 
                ? addressDetails.ElectricityMeters.Select(em => new LookupElectricityMeter(em)).ToList() 
                : null;
            GasMeters = addressDetails.GasMeters != null 
                ? addressDetails.GasMeters.Select(gm => new LookupGasMeter(gm)).ToList() 
                : null;
        }
    }
}